File: Crystal/MathTest/Cone3dTest.cpp
#include "gtest/gtest.h"

#include "../Math/Cone3d.h"

using namespace Crystal::Math;

namespace {
	constexpr auto tolerance = 1.0e-12;
}

TEST(Cone3dTest, TestGetPositionUV)
{
	const Cone3d cone;
	{
		const auto& actual = cone.getPosition(0.0, 0.0);
		const Vector3dd expected(1.0, 0.0, 0.0);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
	{
		const auto& actual = cone.getPosition(0.5, 0.0);
		const Vector3dd expected(-1.0, 0.0, 0.0);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
	{
		const auto& actual = cone.getPosition(0.0, 0.5);
		const Vector3dd expected(0.5, 0.0, 0.5);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
	{
		const auto& actual = cone.getPosition(0.5, 0.5);
		const Vector3dd expected(-0.5, 0.0, 0.5);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
	{
		const auto& actual = cone.getPosition(1.0, 1.0);
		const Vector3dd expected(0.0, 0.0, 1.0);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
	{
		const auto& actual = cone.getPosition(0.0, 1.0);
		const Vector3dd expected(0.0, 0.0, 1.0);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
}

TEST(Cone3dTest, TestGetPositionUVW)
{
	const Cone3d cone;
	{
		const auto& actual = cone.getPosition(0.0, 0.0, 1.0);
		const Vector3dd expected(1.0, 0.0, 0.0);
		EXPECT_TRUE( areSame(expected, actual, tolerance) );
	}
	{
		const auto& actual = cone.getPosition(0.5, 0.0, 1.0);
		const Vector3dd expected(-1.0, 0.0, 0.0);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
	{
		const auto& actual = cone.getPosition(0.0, 0.5, 1.0);
		const Vector3dd expected(0.5, 0.0, 0.5);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
	{
		const auto& actual = cone.getPosition(0.5, 0.5, 1.0);
		const Vector3dd expected(-0.5, 0.0, 0.5);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
	{
		const auto& actual = cone.getPosition(1.0, 1.0, 1.0);
		const Vector3dd expected(0.0, 0.0, 1.0);
		EXPECT_TRUE(areSame(expected, actual, tolerance));
	}
}

File: CGStudio/PGCoreTest/Math/Primitive3d/Cone3dTest.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using PG.Core.Math;

namespace PGCoreTest.Math.Primitive3d
{
    [TestClass]
    public class Cone3dTest
    {
        private readonly double tolerance = 1.0e-12;

        [TestMethod]
        public void TestToCircleAt()
        {
            var cone = new Cone3d(10.0, 100.0, new Vector3d(0, 0, 0));

            {
                var actual = cone.ToCircleAt(0.0);
                var expected = new Circle3d(10.0, new Vector3d(0, 0, 0));
                Assert.IsTrue(expected.IsSame(actual, tolerance));
            }

            {
                var actual = cone.ToCircleAt(0.1);
                var expected = new Circle3d(9.0, new Vector3d(0, 0, 10));
                Assert.IsTrue(expected.IsSame(actual, tolerance));
            }

            {
                var actual = cone.ToCircleAt(1.0);
                var expected = new Circle3d(0.0, new Vector3d(0, 0, 100));
                Assert.IsTrue(expected.IsSame(actual, tolerance));
            }
        }

        [TestMethod]
        public void TestGetPosition()
        {
            var cone = new Cone3d(10.0, 100.0, new Vector3d(0, 0, 0));

            {
                var actual = cone.GetPosition(1.0, 1.0, 1.0);
                var expected = new Vector3d(0, 0, 100);
                Assert.IsTrue(expected.IsSame(actual, tolerance));
            }
        }
    }
}